syntax = "proto3";

package comm;

message PolarPosition {
    //use the system of polar coordinates
    double distance=1;
    double angle=2;
}

message XYPosition {
    //use the system of rectangular coordinates
    double x=1;
    double y=2;
}

message Item {
    //items like gun, etc.
    int32 item_ID=1;
    int32 type=2;
    PolarPosition pos=3;
    int32 durability=4;
}

message SelfInfo {
    //complete info of self
    double heal_point=1;
    double heal_point_limit=2;
    double move_angle=3;
    double view_angle=4;
    double move_speed=5;
    int32 vocation=6;
    int32 status=7;
    int32 move_cd=8;
	int32 attack_cd=9;
    XYPosition pos=10;
    repeated Item bag=11;
	double view_width = 12;
	double view_distance = 13;
}

message OtherInfo {
    //basic info of other characters in visual field
    int32 player_ID=1;
    int32 status=2;
    double move_angle=3;
    double view_angle=4;
    double move_speed=5;
	int32 vocation=6;
    PolarPosition pos=7;
}

message Sound {
    int32 sender=1;
    int32 delay=2;
    int32 parameter=3;
}

message PlayerInfo {
    //all info that send to each player per turn
    int32 player_ID=6;
    SelfInfo self=1;
    repeated int32 landform_id=2;
    repeated Item items=3;
    repeated OtherInfo others=4;
    repeated Sound sounds=5;
}

enum CommandType {
    MOVE=0;
    SHOOT=1;
    PICKUP=2;
    RADIO=3;
    PARACHUTE=4;
}

message Command {
    CommandType command_type=1;
    int32 target_ID=2;
    double move_angle=3;
    double view_angle=4;
    int32 parameter=5;
    int32 role=6;
    XYPosition landing_point=7;
}

message Route {
    XYPosition start_pos=1;
    XYPosition over_pos=2;
    repeated int32 teammates=3;
}
